File: it-agents/skills/code-architect/references/code-examples/csharp/Strategy.cs
// Strategy — a family of interchangeable algorithms behind one interface.
using System;
using System.Collections.Generic;
using System.Linq;

namespace Patterns.Strategy
{
    interface IStrategy
    {
        IEnumerable<int> Apply(IEnumerable<int> data);
    }

    sealed class AscendingStrategy : IStrategy
    {
        public IEnumerable<int> Apply(IEnumerable<int> data) => data.OrderBy(x => x);
    }

    sealed class DescendingStrategy : IStrategy
    {
        public IEnumerable<int> Apply(IEnumerable<int> data) => data.OrderByDescending(x => x);
    }

    sealed class EvensFirstStrategy : IStrategy
    {
        public IEnumerable<int> Apply(IEnumerable<int> data) => data.OrderBy(x => x % 2 != 0).ThenBy(x => x);
    }

    // Context delegates the algorithm to the configured strategy.
    sealed class Context
    {
        private IStrategy _strategy;
        public Context(IStrategy strategy) => _strategy = strategy;
        public void SetStrategy(IStrategy strategy) => _strategy = strategy;
        public string Run(IEnumerable<int> data) => string.Join(",", _strategy.Apply(data));
    }

    static class Demo
    {
        public static void Run()
        {
            int[] data = { 3, 1, 4, 2 };
            var context = new Context(new AscendingStrategy());
            Console.WriteLine(context.Run(data));
            context.SetStrategy(new DescendingStrategy());
            Console.WriteLine(context.Run(data));
            context.SetStrategy(new EvensFirstStrategy());
            Console.WriteLine(context.Run(data));
        }

        public static void Main() => Run();
    }
}

File: it-agents/skills/code-architect/references/code-examples/swift/Strategy.swift
// Strategy — a family of interchangeable algorithms behind one interface.

protocol SortStrategy {
    func sort(_ input: [Int]) -> [Int]
}

struct AscendingStrategy: SortStrategy {
    func sort(_ input: [Int]) -> [Int] { input.sorted(by: <) }
}

struct DescendingStrategy: SortStrategy {
    func sort(_ input: [Int]) -> [Int] { input.sorted(by: >) }
}

// Context delegates the algorithm to a swappable strategy.
final class Arranger {
    private var strategy: SortStrategy

    init(strategy: SortStrategy) {
        self.strategy = strategy
    }

    func use(_ strategy: SortStrategy) {
        self.strategy = strategy
    }

    func arrange(_ input: [Int]) -> [Int] {
        strategy.sort(input)
    }
}

func runDemo() {
    let data = [3, 1, 4, 1, 5, 9, 2]
    let arranger = Arranger(strategy: AscendingStrategy())
    print("ascending:", arranger.arrange(data))

    arranger.use(DescendingStrategy())
    print("descending:", arranger.arrange(data))
}

runDemo()
